### ISSUE TITLE
Run dist 2 gpus , the speed-up radio is low   x1.4

### ISSUE THREAD

**Issue by zwqjoy** [horovod/horovod#1551]
docker image:  horovod:0.18.2-tf2.0.0-torch1.3.0-mxnet1.5.0-py3.6-gpu

Tensorflow2 Test

**Your question:**
I use the  official docker hub image to run tensorflow2 dist test

I run with one process,    the benchmark is Result:  440.6 pic/sec
and I run with 4 processes (1 machine,   2 docker contaniner. )  I expect the result is   440* 2 = 880
but the get  715.9 pic/sec

I do not know why the speed-up radio is so slow

1  
Result:  440.6 pic/sec

```
mpirun --allow-run-as-root -np 1     -H localhost:1,     -mca plm_rsh_args "-p 12345"     -bind-to none -map-by slot     -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH     -x NCCL_SOCKET_IFNAME=^lo,docker0     -mca pml ob1 -mca btl ^openib     -mca btl_tcp_if_exclude lo,docker0     python /home/work/dist/hvd-dist-tf2.py --batch-size-per-replica=128
```

2 
Running benchmark...
Result:  560.9 pic/sec

add compression=hvd.Compression.fp16

Result:  620 pic/sec

```
mpirun --allow-run-as-root -np 2     -H 172.200.22.5:1,172.200.48.8:1    -mca plm_rsh_args "-p 12345"     -bind-to none -map-by slot     -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH     -x NCCL_SOCKET_IFNAME=^lo,docker0     -mca pml ob1 -mca btl ^openib     -mca btl_tcp_if_exclude lo,docker0     python /home/work/dist/hvd-dist-tf2.py --batch-size-per-replica=128
```

[timeline.zip](https://github.com/horovod/horovod/files/3910054/timeline.zip)

**Comment by tgaddair** [comment]
Hey @zwqjoy, thanks for providing the timeline file.  Looks like about half the time is being spent waiting for data, which suggests that we may be able to see some improvement by tuning the fusion buffer threshold and cycle time (try running `horovodrun --autotune` to explore different combinations of these parameters).

I would also suggest running the [tensorflow2_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/tensorflow2_synthetic_benchmark.py) example, which will tell us if the issue is due to your cluster setup or the model.

**Comment by zwqjoy** [comment]
> Hey @zwqjoy, thanks for providing the timeline file. Looks like about half the time is being spent waiting for data, which suggests that we may be able to see some improvement by tuning the fusion buffer threshold and cycle time (try running `horovodrun --autotune` to explore different combinations of these parameters).
> 
> I would also suggest running the [tensorflow2_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/tensorflow2_synthetic_benchmark.py) example, which will tell us if the issue is due to your cluster setup or the model.


Run tensorflow2_synthetic_benchmark.py
1.  one gpu
Img/sec per GPU: 80.8 +-0.7
Total img/sec on 1 GPU(s): 80.8 +-0.7


2  two node, each node one gpu

Img/sec per GPU: 64.0 +-2.6
Total img/sec on 2 GPU(s): 128.1 +-5.3


128/80 == 1.6

**Comment by tgaddair** [comment]
Hey @zwqjoy, the log file should contain the best parameters.  If it's not, then that's a bug and I'll see if I can reproduce the issue.

You can also see them printed to the screen if you run with `horovodrun --log-level INFO ...`.

**Comment by zwqjoy** [comment]
> Hey @zwqjoy, the log file should contain the best parameters. If it's not, then that's a bug and I'll see if I can reproduce the issue.
> 
> One possibility is that the training process is completing before the autotuning finishes its "warmup" period. Can you try running your training script for more steps / epochs?
> 
> You can also see them printed to the screen if you run with `horovodrun --log-level INFO ...`. That will also show you when autotune finishes warming up.
> 
> You can also configure the warmup period with `horovodrun --autotune-warmup-samples`.
@tgaddair  

horovodrun --autotune --autotune-warmup-samples 500 --log-level INFO -np 4 -p 12345 -H 172.200.22.5:2,172.200.48.7:2  python /home/work/dist/hvd-dist-tf2.py --batch-size-per-replica=128

I donot set printed to the screen.
```
1,1]<stdout>:hvd12-1:35057:35285 [1] NCCL INFO comm 0x7fbb1430e920 rank 1 nranks 4 cudaDev 1 nvmlDev 1 - Init COMPLETE
[1,2]<stdout>:hvd13-1:31189:31417 [0] NCCL INFO comm 0x7fc20c30e160 rank 2 nranks 4 cudaDev 0 nvmlDev 0 - Init COMPLETE
[1,3]<stdout>:hvd13-1:31190:31420 [1] NCCL INFO comm 0x7fb7bc316220 rank 3 nranks 4 cudaDev 1 nvmlDev 3 - Init COMPLETE
[1,0]<stdout>:hvd12-1:35056:35284 [0] NCCL INFO Launch mode Parallel
[1,0]<stdout>: 1/25 [>.............................] - ETA: 4:05 - loss: 1.6050 - sparse_categorical_accuracy: 0.2500[1,1]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.437375). Check your callbacks.
[1,1]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.437375). Check your callbacks.
[1,3]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.432434). Check your callbacks.
[1,3]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.432434). Check your callbacks.
[1,2]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.432457). Check your callbacks.
[1,2]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.432457). Check your callbacks.
[1,0]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.434043). Check your callbacks.
[1,0]<stderr>:WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (1.434043). Check your callbacks.
25/25 [==============================] - 21s 843ms/step - loss: 2.8563 - sparse_categorical_accuracy: 0.2588tdout>:[1,0]<stdout>:::
[1,0]<stdout>:Epoch 2/5
[1,0]<stdout>:25/25 [==============================] - 11s 444ms/step - loss: 1.2259 - sparse_categorical_accuracy: 0.4436tdout>:
[1,0]<stdout>:Epoch 3/5
[1,0]<stdout>:[2019-12-06 17:32:28.148261: I horovod/common/parameter_manager.cc:181] Autotuner: Warming up (499 remaining)
[1,0]<stdout>:25/25 [==============================] - 11s 434ms/step - loss: 0.9328 - sparse_categorical_accuracy: 0.6281tdout>::
[1,0]<stdout>:Epoch 4/5
25/25 [==============================] - 11s 430ms/step - loss: 0.6347 - sparse_categorical_accuracy: 0.7432tdout>:[1,0]<stdout>::
[1,0]<stdout>:Epoch 5/5
 1/25 [>.............................] - ETA: 10s - loss: 0.4674 - sparse_categorical_accuracy: 0.8203[1,0]<stdout>:[2019-12-06 17:32:50.172974: I horovod/common/parameter_manager.cc:181] Autotuner: Warming up (498 remaining)
25/25 [==============================] - 11s 428ms/step - loss: 0.3560 - sparse_categorical_accuracy: 0.8516 0.8488[1,0]<stdout>::
[1,0]<stdout>:Running benchmark...
[1,0]<stdout>:per_epoch_time: 43.39752435684204
[1,0]<stdout>:Result:  1179.8 pic/sec
```
